### PR TITLE
Add syntax in try/catch to retrieve the stacktrace directly

### PR DESCRIPTION
**Pull request by bjorng**
This PR adds a new syntax for retrieving the stacktrace without calling `erlang:get_stacktrace/0`. That allow us to deprecate `erlang:get_stacktrace/0` and ultimately remove it.

The problem with `erlang:get_stacktrace/0` is that it can keep huge terms in a process for an indefinite time after an exception. The stacktrace can be huge after a `function_clause` exception or a failed call to a BIF or operator, because the arguments for the call will be included in the stacktrace. For example:

```
1> catch abs(lists:seq(1, 1000)).
{'EXIT',{badarg,[{erlang,abs,
                         [[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20|...]],
                         []},
                 {erl_eval,do_apply,6,[{file,"erl_eval.erl"},{line,674}]},
                 {erl_eval,expr,5,[{file,"erl_eval.erl"},{line,431}]},
                 {shell,exprs,7,[{file,"shell.erl"},{line,687}]},
                 {shell,eval_exprs,7,[{file,"shell.erl"},{line,642}]},
                 {shell,eval_loop,3,[{file,"shell.erl"},{line,627}]}]}}
2> erlang:get_stacktrace().
[{erlang,abs,
         [[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,
           23,24|...]],
         []},
 {erl_eval,do_apply,6,[{file,"erl_eval.erl"},{line,674}]},
 {erl_eval,expr,5,[{file,"erl_eval.erl"},{line,431}]},
 {shell,exprs,7,[{file,"shell.erl"},{line,687}]},
 {shell,eval_exprs,7,[{file,"shell.erl"},{line,642}]},
 {shell,eval_loop,3,[{file,"shell.erl"},{line,627}]}]
3>
```

We can extend the syntax for clauses in try/catch to optionally bind the stacktrace to a variable.

Here is an example using the current syntax:

    try
      Expr
    catch C:E ->
      Stk = erlang:get_stacktrace(),
      .
      .
      .

In the new syntax, it would look like:

    try
      Expr
    catch
      C:E:Stk ->
       .
       .
       .
Only a variable (not a pattern) is allowed in the stacktrace position, to discourage matching of the stacktrace. (Matching would also be expensive, because the raw format of the stacktrace would have to be converted to the cooked form before matching.)

Note that:

    try
      Expr
    catch E ->
      .
      .
      .

is a shorthand for:

    try
      Expr
    catch throw:E ->
      .
      .
      .
If the stacktrace is to be retrieved for a throw, the `throw:`
prefix must be explicitly included:

    try
      Expr
    catch throw:E:Stk ->
      .
      .
      .

### CODE REVIEW

**Comment by leoliu** [comment]
Instead of C:R:S, why not {C,R,S}? C:R can be kept as syntax sugar for {C,R}.

**Comment by bjorng** [comment]
@richcarl @margnus1 @mikpe @kostis Could you have a look at the HiPE part of this PR? I have compiled `trycatch_SUITE` to native code and confirmed that the two new test cases succeed, so I know that the basics work.

**Comment by bjorng** [comment]
> Instead of C:R:S, why not {C,R,S}? C:R can be kept as syntax sugar for {C,R}.

Why `{C,R,S}`? What would be the advantage compared to C:R:S?

Also, `{C,R,S}` is ambiguous.  Does it mean `throw:{C,R,S}` or `C:R:S`?

**Comment by leoliu** [comment]
I mean {Class, Reason, Stacktrace} where Class is exit, error or throw. It is a familiar pattern compared to Class:Reason:Stacktrace.

**Comment by bjorng** [comment]
> I mean {Class, Reason, Stacktrace} where Class is exit, error or throw. It is a familiar pattern compared to Class:Reason:Stacktrace.

I got what you meant. I would say say that since we already have the syntax `Class:Reason`, `Class:Reason:Stacktrace` is just as familiar.

What I meant about it being ambiguous is that your suggested syntax already has a meaning, and that existing code would change meaning if your suggestion were to be implemented. Here is an example:

```
Eshell V9.0  (abort with ^G)
1> try throw({1,2,3}) catch {C,R,S} -> C+R+S end.
6
2>
```

**Comment by leoliu** [comment]
Now I see the suggestion isn’t compatible with existing uses. C:R / C:R:S isn’t used anywhere, so it always look odd to me. I was looking for opportunities to get rid of it.

File: erts/emulator/hipe/hipe_bif_list.m4 [hunk 1]
```diff
@@ -186,6 +186,8 @@ gc_bif_interface_1(nbif_erase_1, erase_1)
 gc_bif_interface_1(nbif_erts_internal_garbage_collect_1, erts_internal_garbage_collect_1)
 gc_nofail_primop_interface_1(nbif_gc_1, hipe_gc)
 gc_bif_interface_2(nbif_put_2, put_2)
+gc_bif_interface_2(nbif_put_2, put_2)
```
[the remark this answers is not in the document]
Comment (bjorng) [reply erts/emulator/hipe/hipe_bif_list.m4]: Yes, fixed!

**Comment by mikpe** [comment]
HiPE bits LGTM.

**Comment by uabboli** [comment]
erl_parse.yrl looks OK.

It seems that syntax_tools needs to be updated. For example, erl_prettypr skips the stacktrace variable.

**Comment by richcarl** [comment]
Will you try to optimize the case when someone does a rethrow and never looks at the contents of the variable (or passes it outside the scope in any other way than as the stacktrace in a rethrow)? In such cases, the raw internal value doesn't have to be blown up into an actual list-of-tuples.

**Comment by bjorng** [comment]
> Will you try to optimize the case when someone does a rethrow and never looks at the contents of the variable (or passes it outside the scope in any other way than as the stacktrace in a rethrow)? In such cases, the raw internal value doesn't have to be blown up into an actual list-of-tuples.

Good idea! I will try to do that, at least before OTP 21 is released.

**Comment by bjorng** [comment]
I have added a fixup commit that will produce a compilation error if the stacktrace variable is previously bound. For example, the following code:

```
foo(E) ->
    Stk = some_value,
    try
        E()
    catch
        C:E:Stk ->
            {C,E,Stk}
    end.
```

will cause the following compilation error to be issued:

`some_file.erl:9: stacktrace variable 'Stk' must not be previously bound
`

**Comment by vans163** [comment]
Is it possible to allow pattern match on the stacktrace fully knowing your taking a perf penalty?

**Comment by brigadier** [comment]
Am I right that even after this PR stacktrace will be built until the `get_stacktrace` is removed? Maybe it is possible to eliminate stacktrace building earlier, with a compile time option?

**Comment by bjorng** [comment]
> Am I right that even after this PR stacktrace will be built until the get_stacktrace is removed?

No stacktrace is built when an exception occurs, but information is saved so that the stacktrace can be built later when requested. The information is quite small except for the arguments for a `function_clause` or `badarg` exception when calling a BIF.

After this patch, the information will still be saved indefinitely in the process. When we'll change `stacktrace/0` to return an empty list in some future release, the information no longer needs to be saved in the process.

> Maybe it is possible to eliminate stacktrace building earlier, with a compile time option for E/OTP?

No.

**Pull request merged**